system/iptables.py

```diff
@@ -255,8 +249,16 @@ def append_match(rule, param, match):
 def construct_rule(params):
     rule = []
     append_param(rule, params['protocol'], '-p', False)
-    append_param(rule, params['source'], '-s', False)
-    append_param(rule, params['destination'], '-d', False)
+    source = params['source']
+    if (source is not None) and (len(source)>0) and (source[0]=='!'):
+        rule.extent(['!'])
+        source = source[1:]
+    append_param(rule, source, '-s', False)
+    destination = params['destination']
+    if (destination is not None) and (len(destination)>0) and (destination[0]=='!'):
```

Comment:
This looks very crammed it should have some whitespace (`missing whitespace around operator`) and I also think that it would be good to extract this into a function.